File: ruby/dominoes/dominoes.rb
class Array
  def without_first(e)
    i = index(e)
    return self if i.nil?
    r = dup
    r.delete_at(i)
    r
  end
end

module Dominoes
  def self.chain?(dominoes)
    chains([], dominoes).any? do |chain|
      chain.empty? || chain.first.first == chain.last.last
    end
  end

  private

  def self.chains(chain, dominoes)
    return [chain] if dominoes.empty?
    dominoes.flat_map do |domino|
      if chain.empty?
        ch = [[domino]]
        if domino.first != domino.last
          ch << [domino.reverse]
        end
        ch.flat_map { |c| chains(c, dominoes.without_first(domino)) }
      elsif chain.last.last == domino.first
        chains(chain + [domino], dominoes.without_first(domino))
      elsif chain.last.last == domino.last
        chains(chain + [domino.reverse], dominoes.without_first(domino))
      else
        []
      end
    end
  end
end

File: clojure/dominoes/src/dominoes.clj
(ns dominoes)

(defn- remove-first [x coll]
  (let [[before [_ & after]] (split-with (partial not= x) coll)]
    (concat before after)))

(defn- reverse-domino [domino]
  (vec (rseq domino)))

(defn- chain-head [[[head]]]
  head)

(defn- chain-tail [chain]
  (last (last chain)))

(defn- valid-chain? [chain]
  (= (chain-head chain) (chain-tail chain)))

(defn- chains [dominoes chain]
 (if (empty? dominoes)
   [chain]
   (mapcat
     (fn [domino]
       (let [tail (chain-tail chain)
             remaining-dominoes (remove-first domino dominoes)]
         (cond
           (= (first domino) tail) (chains remaining-dominoes (conj chain domino))
           (= (last domino) tail)  (chains remaining-dominoes (conj chain (reverse-domino domino)))
           :else                   [])))
     dominoes)))

(defn can-chain? [[first-domino & other-dominoes]]
  (or
    (nil? first-domino)
    (some valid-chain? (chains other-dominoes [first-domino]))))
